lib/bolt/executor.rb

```diff
@@ -373,83 +374,54 @@ def run_plan(scope, plan, params)
       plan.call_by_name_with_scope(scope, params, true)
     end
 
-    def create_yarn(scope, block, object, index)
-      fiber = Fiber.new do
-        # Create the new scope
-        newscope = Puppet::Parser::Scope.new(scope.compiler)
-        local = Puppet::Parser::Scope::LocalScope.new
-
-        # Compress the current scopes into a single vars hash to add to the new scope
-        current_scope = scope.effective_symtable(true)
-        until current_scope.nil?
-          current_scope.instance_variable_get(:@symbols)&.each_pair { |k, v| local[k] = v }
-          current_scope = current_scope.parent
-        end
-        newscope.push_ephemerals([local])
+    # Call into FiberExecutor to avoid this class getting
+    # overloaded while also minimizing the Puppet lookups needed from plan
+    # functions
+    #
+    def create_future(scope, block, name = nil)
+      @fiber_executor.create_future(scope, block, name)
+    end
 
-        begin
-          result = catch(:return) do
-            args = { block.parameters[0][1].to_s => object }
-            block.closure.call_by_name_with_scope(newscope, args, true)
-          end
+    def add_future(future)
+      @fiber_executor.add_future(future)
+    end
 
-          # If we got a return from the block, get it's value
-          # Otherwise the result is the last line from the block
-          result = result.value if result.is_a?(Puppet::Pops::Evaluator::Return)
+    def plan_complete?
+      @fiber_executor.plan_complete?
+    end
 
-          # Validate the result is a PlanResult
-          unless Puppet::Pops::Types::TypeParser.singleton.parse('Boltlib::PlanResult').instance?(result)
-            raise Bolt::InvalidParallelResult.new(result.to_s, *Puppet::Pops::PuppetStack.top_of_stack)
-          end
+    def round_robin
+      @fiber_executor.round_robin
+    end
 
-          result
-        rescue Puppet::PreformattedError => e
-          if e.cause.is_a?(Bolt::Error)
-            e.cause
-          else
-            raise e
-          end
-        end
-      end
+    def in_parallel
+      @fiber_executor.in_parallel
+    end
 
-      Bolt::Yarn.new(fiber, index)
+    def wait(futures, timeout, **opts)
+      @fiber_executor.wait(futures, timeout, **opts)
     end
 
-    def handle_event(event)
-      case event[:type]
-      when :node_result
-        @thread_completed = true
-      end
+    def plan_futures
+      @fiber_executor.plan_futures
     end
 
-    def round_robin(skein)
-      subscribe(self, [:node_result])
-      results = Array.new(skein.length)
-      @in_parallel = true
-      publish_event(type: :stop_spin)
+    def run_in_thread(cmd, params)
```

Comment:
I think this would be easier to understand/more idiomatic if it accepted a block rather than a method name and partial params. We already handling the stack in the boltlib functions for the non-parallel case so abstracting it here makes it harder to understand. 
A block accepting function would also be more general purpose and could be implemented on the `fiber_executor` perhaps.

Reply:
100%, not sure why I didn't think of that.